### PR TITLE
Add trait for serialization

### PR DESCRIPTION
**Pull request by kampersanda**
This PR added traits for serializing/deserializing data structures to simplify codes.

### CODE REVIEW

File: src/bytewise.rs [hunk 1]
```diff
@@ -11,6 +11,9 @@ use alloc::vec::Vec;
 use crate::build_helper::BuildHelper;
 use crate::errors::{DaachorseError, Result};
 use crate::intpack::{U24nU8, U24};
+use crate::serializer::{
+    deserialize_vec, serialize_slice, serialized_bytes, Deserialize, Serialize,
```
Comment (vbkaisetsu) [inline src/bytewise.rs:15]: Could you follow the idiomatic way to bring fuctions?
https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html#creating-idiomatic-use-paths
> Specifying the parent module when calling the function makes it clear that the function isn’t locally defined while still minimizing repetition of the full path.

**Review by vbkaisetsu — COMMENTED** [review]
I think the API should be as follows.

* Merge `Serialize` and `Deserialize` into the `Serializable` trait.
* Rename `to_vec` to `serialize_to_vec` because such name is already used in the slice and could be confusing.
* Remove the glogal `serialize_slice()` function, and implement `Serializable` in `Vec<S>`.
* Implement `serialized_bytes()` as the function of `Serializable`.

**Comment by kampersanda** [comment]
@vbkaisetsu Thank you for your suggestions. I've modified codes following them.
I also implemented `Serializable` for `MatchKind` and `CodeMapper` for consistency.
Could you review them again?

File: src/serializer.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            let (x, rest) = S::deserialize_from_slice(src);
+            dst.push(x);
+            src = rest;
+        }
+        (dst, src)
+    }
+
+    fn serialized_bytes(&self) -> usize {
+        mem::size_of::<u32>() + mem::size_of::<S>() * self.len()
+    }
```
Comment (vbkaisetsu) [inline src/serializer.rs:71]: This is incorrect
```rust
let v1 = vec![0u32; 10];
dbg!(v1.serialized_bytes());
let v2 = vec![v1; 10];
dbg!(v2.serialized_bytes());
```
```
v1.serialized_bytes() = 44
v2.serialized_bytes() = 244
```

File: src/serializer.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+use crate::utils::FromU32;
+
+pub trait Serializable: Sized {
+    fn serialize_to_vec(&self, dst: &mut Vec<u8>);
+
+    fn deserialize_from_slice(src: &[u8]) -> (Self, &[u8]);
+
+    fn serialized_bytes(&self) -> usize {
+        mem::size_of::<Self>()
+    }
```
Comment (vbkaisetsu) [inline src/serializer.rs:17]: `size_of()` depends on its layout.

**Comment by kampersanda** [comment]
@vbkaisetsu I fixed as follows:

- Specialized `Serializable` for `Vec<S>` as `SerializableVec`.
- Added tests
- Implemented `Serializable::serialized_bytes` in each specialization without shared behavior.

File: src/lib.rs [hunk 4]
```diff
@@ -343,3 +358,41 @@ impl From<MatchKind> for u8 {
...
+    }
+
+    #[inline(always)]
+    fn deserialize_from_slice(src: &[u8]) -> (Self, &[u8]) {
+        (Self::from(src[0]), &src[1..])
+    }
+
+    #[inline(always)]
+    fn serialized_bytes() -> usize {
+        core::mem::size_of::<Self>()
```
Comment (vbkaisetsu) [inline src/lib.rs:375]: ditto

File: src/bytewise.rs [hunk 5]
```diff
@@ -784,23 +754,34 @@ impl State {
...
+                fail,
+                opos_ch,
+            },
+            src,
+        )
+    }
+
+    #[inline(always)]
+    fn serialized_bytes() -> usize {
+        mem::size_of::<Self>()
```
Comment (vbkaisetsu) [inline src/bytewise.rs:784]: ditto

**Comment by kampersanda** [comment]
@vbkaisetsu To keep consistency to the implementations of `serialize_to_vec` and `deserialize_from_slice`, I fixed `serialized_bytes` to use `Serializable::serialized_bytes()`.

File: src/bytewise.rs [hunk 6]
```diff
@@ -994,4 +977,29 @@ mod tests {
         assert_eq!(0x80, pma.states[0].base().unwrap().get());
         assert_eq!(0x100, pma.states[0x80].base().unwrap().get());
     }
+
+    #[test]
+    fn test_serialize_state() {
+        let mut opos_ch = U24nU8::default();
+        opos_ch.set_a(U24::try_from(57).unwrap());
+        opos_ch.set_b(77);
+        crate::serializer::tests::test_common(State {
```
Comment (vbkaisetsu) [inline src/bytewise.rs:986]: Could you not hide the test code inside the function?

In unit testing, it is often said DAMP not DRY.
It is not clear what the behavior of `test_common` is, so we need to move on to the definition of this function.
https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests

Comment (kampersanda) [reply src/bytewise.rs]: I removed `test_common` https://github.com/daac-tools/daachorse/pull/68/commits/32618fa25af522d9bb9aaece0f6923a135e0352f
I learned a lot, thanks!

File: src/serializer.rs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    use super::*;
+
+    #[test]
+    fn test_u32() {
+        let x = 42u32;
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), u32::serialized_bytes());
+        let (y, rest) = u32::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
Comment (vbkaisetsu) [inline src/serializer.rs:100]:
```suggestion
        let x = 0x01234567u32;
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(vec![0x67, 0x45, 0x23, 0x01], data);
        assert_eq!(4, u32::serialized_bytes());
        data.push(42);
        let (y, rest) = u32::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

File: src/serializer.rs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    #[test]
+    fn test_nzu32() {
+        let x = NonZeroU32::new(42u32);
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), Option::<NonZeroU32>::serialized_bytes());
+        let (y, rest) = Option::<NonZeroU32>::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
Comment (vbkaisetsu) [inline src/serializer.rs:111]:
```suggestion
        let x = NonZeroU32::new(0x01234567u32);
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(vec![0x67, 0x45, 0x23, 0x01], data);
        assert_eq!(4, Option::<NonZeroU32>::serialized_bytes());
        data.push(42);
        let (y, rest) = Option::<NonZeroU32>::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

File: src/serializer.rs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    #[test]
+    fn test_vec_u32() {
+        let x = vec![42u32; 10];
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), x.serialized_bytes());
+        let (y, rest) = Vec::<u32>::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
Comment (vbkaisetsu) [inline src/serializer.rs:122]:
```suggestion
        let x = vec![0x01234567u32, 0x89abcdefu32, 0x02468aceu32];
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(
            vec![
                0x03, 0x00, 0x00, 0x00, // len
                0x67, 0x45, 0x23, 0x01, // item 1
                0xef, 0xcd, 0xab, 0x89, // item 2
                0xce, 0x8a, 0x46, 0x02, // item 3
            ],
            data
        );
        assert_eq!(16, x.serialized_bytes());
        data.push(42);
        let (y, rest) = Vec::<u32>::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

**Review by vbkaisetsu — COMMENTED** [review]
I suggest adding verification of serialized data because it is small and easy to check them.

**Review by kg86 — APPROVED** [review]

**Pull request merged**